cvmfs/cache_plugin/cvmfs_cache_posix.cc

```diff
@@ -0,0 +1,415 @@
+  if (debug_log != NULL) {
+    SetLogDebugFile(debug_log);
+    cvmcache_options_free(debug_log);
+  } else {
+    SetLogDebugFile("/dev/null");
+  }
+  char *locator = cvmcache_options_get(options, "CVMFS_CACHE_PLUGIN_LOCATOR");
+  if (locator == NULL) {
+    LogCvmfs(kLogCache, kLogStderr | kLogSyslogErr,
+             "CVMFS_CACHE_PLUGIN_LOCATOR missing");
+    cvmcache_options_fini(options);
+    return 1;
+  }
+  char *cache_path = cvmcache_options_get(options, "CVMFS_CACHE_PLUGIN_PATH");
+  if (cache_path == NULL) {
+    LogCvmfs(kLogCache, kLogStderr | kLogSyslogErr,
+             "CVMFS_CACHE_PLUGIN_PATH missing");
+    cvmcache_options_free(locator);
+    cvmcache_options_fini(options);
+    return 1;
+  }
+  char *test_mode = cvmcache_options_get(options, "CVMFS_CACHE_PLUGIN_TEST");
+  if (!test_mode) {
+    char *watchdog_crash_dump_path =
+      cvmcache_options_get(options, "CVMFS_CACHE_PLUGIN_CRASH_DUMP");
+    cvmcache_spawn_watchdog(watchdog_crash_dump_path);
+    if (watchdog_crash_dump_path)
+      cvmcache_options_free(watchdog_crash_dump_path);
+  }
+
+  if (!MkdirDeep(cache_path, 0755)) {
+    LogCvmfs(kLogCache, kLogStderr | kLogSyslogErr,
+             "Cannot create cache directory %s", cache_path);
+    return 1;
+  }
+  LogCvmfs(kLogCache, kLogDebug, "Created cache directory %s", cache_path);
+
+  cache_mgr = PosixCacheManager::Create(cache_path, true /* alien cache */);
+
+  cvmcache_hash empty_hash;
+  empty_hash.algorithm = 0;
+  memset(empty_hash.digest, 0, 20);
+  opened_objects = new SmallHashDynamic<cvmcache_hash, CacheObject>;
+  transactions = new SmallHashDynamic<uint64_t, Txn>;
+  listings = new SmallHashDynamic<uint64_t, Listing>;
+  opened_objects->Init(32, empty_hash, cvmcache_hash_hasher);
+  transactions->Init(32, INT64_MAX, uint64_hasher);
+  listings->Init(32, INT64_MAX, uint64_hasher);
+  pinned_size = 0;
+  used_size = 0;
+  capacity = CVMCACHE_SIZE_UNKNOWN;
+
+  struct cvmcache_callbacks callbacks;
+  memset(&callbacks, 0, sizeof(callbacks));
+  callbacks.cvmcache_chrefcnt = posix_chrefcnt;
+  callbacks.cvmcache_obj_info = posix_obj_info;
+  callbacks.cvmcache_pread = posix_pread;
+  callbacks.cvmcache_start_txn = posix_start_txn;
+  callbacks.cvmcache_write_txn = posix_write_txn;
+  callbacks.cvmcache_commit_txn = posix_commit_txn;
+  callbacks.cvmcache_abort_txn = posix_abort_txn;
+  callbacks.cvmcache_info = posix_info;
+  callbacks.cvmcache_breadcrumb_store = posix_breadcrumb_store;
+  callbacks.cvmcache_breadcrumb_load = posix_breadcrumb_load;
+  callbacks.capabilities = CVMCACHE_CAP_WRITE + CVMCACHE_CAP_REFCOUNT +
+                           CVMCACHE_CAP_INFO + CVMCACHE_CAP_BREADCRUMB;
+
+  ctx = cvmcache_init(&callbacks);
+  int retval = cvmcache_listen(ctx, locator);
+  if (!retval) {
+    LogCvmfs(kLogCache, kLogStderr | kLogSyslogErr,
+             "failed to listen on %s", locator);
+    return 1;
+  }
+
+  if (test_mode) {
```

Comment:
What exactly would be the test mode?  Foreground?

Reply:
It ensures that the plugin cooperates with https://github.com/cvmfs/cvmfs/blob/devel/ci/run_unittests.sh . I copied this part of code from https://github.com/cvmfs/cvmfs/blob/devel/cvmfs/cache_plugin/cvmfs_cache_ram.cc .

Reply:
Oh, right.